README.rst
.. image:: https://travis-ci.org/collective/plone.recipe.varnish.svg?branch=master
    :target: https://travis-ci.org/collective/plone.recipe.varnish

.. image:: https://coveralls.io/repos/collective/plone.recipe.varnish/badge.svg?branch=master&service=github
    :target: https://coveralls.io/github/collective/plone.recipe.varnish?branch=master

Varnish recipe for buildout
===========================

plone.recipe.varnish is a `zc.buildout`_ recipe to install `Varnish`_. Even
though the name contains the name Plone, there is nothing Plone-specific about
this recipe: it works for non-Zope sites just as well.

Configuring it is very simple. For example::

    [varnish-build]
    recipe = plone.recipe.varnish:build

    [varnish-configuration]
    recipe = plone.recipe.varnish:configuration
    backends = 127.0.0.1:8081

    [varnish-script]
    recipe = plone.recipe.varnish:script
    bind = 127.0.0.1:8000
    cache-size = 512M


This configures three buildout parts:

``varnish-build``
    which will download, compile and install varnish,

``varnish-configuration``
    which generates the VCL configuration file,
    sending requests to a backend at 127.0.0.1:8081, and

``varnish-script``
    which runs Varnish, configured to listen on 127.0.0.1:8000 for requests,
    using a 512 megabyte cache.

A wrapper script for the varnish startup command is created in the ``bin``
directory of your buildout.

Please note that the configuration generated by this recipe requires Varnish
4.0.x or later. The URL provided by the recipe points to the
latest tested Varnish release compatible with the generated configuration.


Virtual hosting
---------------

Varnish supports virtual hosting by selecting a different backend server
based on headers on the incoming request. You can configure the backends
through the backends option::

  [varnish-configuration]
  backends =
     plone.org:127.0.0.1:8000
     plone.net:127.0.0.1:9000

This will generate a configuration which sends all traffic for the plone.org
host to a backend server running on port 8000 while all traffic for the
plone.net host is send to port 9000.


Zope 2 hosting (with Virtual Host Monster)
------------------------------------------

If you are using Zope 2 as backend server you will need to rewrite the URL
so the Zope Virtual Host Monster (VHM) can generate correct links for links in
your pages. This can be done either by a web server such as Apache or nginx
(placed either in front or behind Varnish) but can also be done by Varnish itself.

The three options are described below.

Option 1 (rewrites after Varnish)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If generating these VHM-style URLs in a proxy *behind* Varnish (or if using
VHM's 'mapping' feature), no extra Varnish configuration is needed.
Just make sure the ``backends`` option directs the traffic to the proxy.

Option 2 (rewrites before Varnish)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

If generating these VHM-style URLs in a proxy *in front* of Varnish, no extra
Varnish configuration is needed as long as the original hostname is still retained
in the URL. If the hostname is not retained, you can tell Varnish to direct requests
based on the "path" instead of the hostname.  For example::

  [varnish-configuration]
  backends =
    /VirtualHostBase/http/plone.org:80/Plone:127.0.0.1:8000
    /VirtualHostBase/http/plone.net:80/Plone:127.0.0.1:9000

This will generate a configuration which sends all traffic for any request whose
path starts with ``/VirtualHostBase/http/plone.org:80/Plone`` to a backend server
running at 127.0.0.1 on port 8000, while request paths starting with
``/VirtualHostBase/http/plone.net:80/Plone`` are sent to port 9000.

Option 3 (rewrites within Varnish)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To have Varnish generate these VHM-style URLs, you can use the
``zope2_vhm_map`` option.
Here is an example::

  [varnish-configuration]
  zope2_vhm_map =
      plone.org:/plone
      plone.net:/plone

This tells us that the domain plone.org should be mapped to the location
``/plone`` in the backend. By combining this with the information from the
``backends`` option a varnish configuration will be generated that
maps URLs correctly.

Load Balancing
--------------

Varnish supports load balancing by configuring a director for a pool of backends.
This director sends the incoming requests that cannot be fulfilled by varnish to
backends in the pool in either random or round robin fashion. You can configure
the director via the ``balancer`` option::

  [varnish-configuration]
  balancer = random

This will generate a configuration which sends all traffic to the director,
which will choose a 'random' backend server to fulfill the request if the
content requested is not cached by varnish itself.


plone.recipe.varnish reference
------------------------------

The ``plone.recipe.varnish`` recipe does one or more of the following:

``plone.recipe.varnish:build``
    compiles varnish from sources

``plone.recipe.varnish:configuration``
    generates a VCL-configuration file

``plone.recipe.varnish:script``
    generates a wrapper script inside your buildout that will start Varnish
    with the correct configuration.

Please note that this recipe requires Varnish 4.0.x or later.


Build varnish from sources
~~~~~~~~~~~~~~~~~~~~~~~~~~

``build`` is based on
`zc.recipe.cmmi <https://pypi.python.org/pypi/zc.recipe.cmmi>`_ - so all
parameters from that recipe are available here too (but rarely used). These options are available for the recipe part plone.recipe.varnish:build.

Two parameters are different/ extra:

``url``
    Location used for download of varnish sources. Defaults to a version tested
    for the selected ``varnish_version``.

``jobs``
    Passes the number of parallel jobs to ``make``, defaults to ``4``. Adjust as
    needed to your CPU resources.

``varnish_version``
    Varnish target version. Default is ``4``.  Options are:

    - 4.0: uses 4.0.3, will stick to 4.0.x
    - 4.1: uses 4.1.2, will stick to 4.1.x
    - 4: uses 4.1.2, will stick to 4.x


VCL Configuration Generator
~~~~~~~~~~~~~~~~~~~~~~~~~~~

These options are available for the recipe part plone.recipe.varnish:configuration.

``backends``
    Specifies the backend or backends which will process the (uncached)
    requests. The syntax for backends:

    ``[<hostname>][/<path>]:<ip address>:<port>``

    The optional ``hostname`` and ``path`` allows you to do virtual hosting.
    If multiple backends are specified then each backend must include
    either a hostname or path (or both) so that Varnish can direct the
    matching request to the appropriate backend. Defaults to
    ``127.0.0.1:8080``.

``balancer``
    If included and set to either ``random`` or ``round-robin``, this option
    configures varnish to load balance the servers specified by the ``backends``
    directive. Possible values: ``none`` (default), ``round-robin`` or
    ``random``.

``between-bytes-timeout``
    If specified, this option configures the timeout (in seconds) for Varnish
    waiting between bytes when receiving data from a backend. Varnish will only
    wait this many seconds between bytes before giving up. A value of 0s means
    this will never time out. Defaults to *60s*, as per Varnish's default
    settings.

``bind``
    Hostname and port on which Varnish will listen for requests. Defaults
    to ``127.0.0.1:8000``.

``connect-timeout``
    If specified, this option configures the connection timeout (in seconds)
    for Varnish connecting to a backend server. Varnish will only try to
    connect to a given backend for this many seconds before giving up. Defaults
    to 0.4s, as per Varnish's default settings.

``cookie-whitelist``
    After the ``cookie-pass`` was processed this list is used to to sanitize
    cookie data on the request. Cookie data to be sent to the backend includes
    only cookies with the given namens. Goal is to work better with the
    backend, i.e. detect if user is logged in and adjust caching to ensure no
    authenticated pages get cached. Defaults are optimized for Zope2/Plone:
    ``statusmessages __ac _ZopeId __cp``

``cookie-pass``
    This list consists of lines with a cookie-match and urlexclude in the form:
    ``"cookiematch":"urlexcludes"``. If *cookiematch* applies for the cookiename
    and the current url does not match urlexcludes, the request is passed
    directly to the configured backend bypassing any caching. But if cookie
    applies and url matches urlexcludes, then a lookup is forced. Defaults are
    optimized for Plone, one line:
    ``"__ac(|_(name|password|persistent))=":"\.(js|css|kss)"``

``first-byte-timeout``
    If specified, this option configures the timeout (in seconds) for Varnish
    receiving the first byte from a backend. Varnish will only wait for this
    many seconds before giving up. A value of 0s means Varnish will never time
    out. Defaults to 300s.

``purge-hosts``
    Specifies hostnames or IP addresses for purge ACL. By default ``localhost`` and
    the backends are allowed to purge. Additional allowed hosts are listed here.

``vcl_recv``, ``vcl_hit``, ``vcl_miss``, ``vcl_fetch``, ``vcl_deliver``, ``vcl_pipe``
    Insert arbitrary VCL code into the generated config.

``verbose-headers``
    Enable sending extra headers in responses that expose what varnish
    did with the request and the cache status. Useful for debugging
    cache settings and optimizations.
    Possible values: ``on`` or ``off`` (default).

``zope2_vhm_map``
    Defines a virtual host mapping for Zope servers. This is a list of
    ``hostname:ZODB location`` entries which specify the location inside
    Zope where the website for a virtual host lives.

``zope2_vhm_port``
    Defines a virtual host mapping port to use in the VHM URL to send back to
    clients. Useful if there is another port mapping in front of varnish, such
    as haproxy. Defaults to bind port.

``zope2_vhm_ssl``
    If specified, this maps VHM URLs to ``https`` for all requests.
    Possible values: ``on`` or ``off`` (default).

``zope2_vhm_ssl_port``
    Defines a virtual host mapping port to use in the VHM URL to send back to
    clients. Useful if there is another port mapping in front of varnish, such
    as haproxy. Defaults to 443.

``varnish_version``
    Varnish target version. If not given,
    it looks for the ``varnish_version`` setting in the build part
    (see the ``build-part`` setting).
    If not given there either, it defaults to ``4``.

To test the generated configuration for syntactic correctness, run
``varnishd -C -f ./parts/varnish-configuration/varnish.vcl``.


Create script to start varnish
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Start varnish as a daemon or in foreground with the given settings. These options are available for the recipe part plone.recipe.varnish:script.

``bind``
    Hostname and port on which Varnish will listen for requests. Defaults
    to ``127.0.0.1:8000``.

``build-part``
    References the buildout part in order to get settings from there. Defaults
    to ``varnish-build``. Set it to ``false`` in order to switch it off.

``cache-location``
    Customise the location for the Varnish file storage.  Option only applicable
    when used with ``file`` or ``persistent`` cache-type options.  Defaults to
    using a file named ``storage`` inside the relevant parts directory
    (eg ``parts/varnish/storage``).  Changing the default location can be
    useful in putting the storage somewhere with quicker read speeds
    (e.g. RAM disk).

``cache-size``
    The size of the cache (limited to 2G on 32bit systems). Defaults to
    256M.

``cache-type``
    Specify the type of cache storage to use with Varnish.
    Possible values: ``file`` (storage for each object is allocated from an
    arena backed by a file),
    ``malloc`` (storage for each object is allocated with malloc; in memory),
    or ``persistent`` (experimental as at Varnish 2.1.4).
    Defaults to ``file``.

``configuration-file``
    Path to a Varnish VCL configuration file to use. Defaults to the generated
    file from the ``configuration-part`` setting.
    If no configuration was generated, this setting is mandatory.

``configuration-part``
    Names the buildout part to get settings from.
    Defaults to ``varnish-configuration``.

``daemon``
    The file and path of the varnish daemon ``varnishd`` to use.
    If not given, it looks for the build part
    (see the ``build-part`` setting)
    and uses its ``location`` setting plus the string ``/sbin/varnishd``.
    If there is no build part, it defaults to ``/usr/sbin/varnishd`` - the
    most common place
    where it's found on many Unix systems. Adjust it if needed.

``grace-healthy``
    Grace in the context of Varnish means delivering otherwise expired objects
    when circumstances call for it. This can happen because:
    (1) the backend-director selected is down, or
    (2) a different thread has already made a request to the backend that's
    not yet finished.

    If the backend is healthy, accept objects that are this number of seconds
    old. Clients will be delivered content that is no more than number of
    seconds past its TTL.

    Format: number followed by a time unit: ms, s, m, h.

    Defaults to ``None``. If this is set to ``None`` the grace
    feature is disabled.

``grace-sick``
    If the backend is sick, accept objects that are this old.
    See also ``grace-healthy``.

    Format: number followed by a time unit: ms, s, m, h.

    Defaults to ``600s``. Should be greater than ``grace-healthy``.

``group``
    The name of the group that varnish should switch to before accepting any
    request. This defaults to the main group for the specified user.

``mode``
    Specify whether the varnish daemon should run in ``daemon`` or
    ``foreground`` mode.  The latter is useful when varnish is run by service
    supervision tools like daemontools or runit. Defaults to ``daemon``.

``name``
    If specified this sets the name of the varnish instance (defaults to
    the host name).

    From varnishd's manpage:

      Amongst other things, this name is used to construct the name of the
      directory in which varnishd keeps temporary files and persistent state.

CHANGES.rst
Changelog
=========

2.0a5 (unreleased)
------------------

- Made three possible values for the ``varnish_version`` option.  4.0
  (uses 4.0.3), 4.1 (uses 4.1.2), 4 (uses 4.1.2).  4 is the default.
  4 is intended to be updated to 4.2.x when this is released and found
  to work.
  [maurits]


2.0a4 (2016-02-23)
------------------

- New: add option for secret-file in the script part so you can communicate to
  varnish with varnishadm. See docs for usage and secret-file generation.
  [fredvd]

- Fix: Split at max on two ':' to get a max of 3 parts as raw_backends
  [jensens]


2.0a3 (2015-12-22)
------------------

- re-release: 2.0a2 was a brown bag release
  [jensens]

2.0a2 (2015-12-22)
------------------

- Fix daemon location of script part of the recipe (/usr/bin/varnishd was
  always used.
  [fredvd]

- Fix tests,  download Varnish 4.0.3 as download.
  [fredvd]

2.0a1 (2015-03-02)
------------------

- refactoring and cleanup of the whole recipe and vcl generation:

  - skip support of varnish < v4.0, use 1.x series for older varnish support.
  - do not generate vcl code in python
  - use jinja2 templates for vcl
  - refactor vcl generation out in own testable class
  - change fixup cookies into a cookie whitelist
  - split up recipe in 3 parts: build, configuration generation and script
    generation.

  [jensens]


1.4 (unreleased)
----------------

- Fix test for running in the Varnish 2 or later.
  [cleberjsantos]

- Fixup VCL template Varnish 3.
  [cleberjsantos]

- add saint-mode for varnish_version 3.
  [cleder, cleberjsantos]

- set a default download-url for varnish_version 3.
  [cleder]

- Fixup string concat for varnish_version 3.
  [damaestro]

- Add zope2_vhm_port to be able to explicitly define a response
  port in VHM URLs.
  [damaestro]

- Add zope2_vhm_ssl to use VHM to render https urls.
  [damaestro]

- Add zope2_vhm_ssl_port to be able to explicitly define a response
  port in VHM URLs for ssl.
  [damaestro]

- Update verbose-headers to use upstream debug example:
  https://www.varnish-cache.org/trac/wiki/VCLExampleHitMissHeader
  [damaestro]

- Add cookie-fixup to better support caching of plone conent
  and to ensure no authenticated content gets cached.
  http://developer.plone.org/hosting/varnish.html
  [damaestro]

- Update VCL templates to be more flexible.
  [damaestro]


1.3 (2013-08-21)
----------------

- Add varnish_version option in order to control vcl generation for
  varnish version >= 3
  [rnix]


1.2.2 (2012-10-14)
------------------

- Moved to https://github.com/collective/plone.recipe.varnish
  [maurits]


1.2.1 (2011-05-13)
------------------

- Update known good Varnish to 2.1.5.
  [elro]

- Add vcl_recv, vcl_hit, vcl_miss, vcl_fetch, vcl_deliver, vcl_pipe options to
  insert arbitrary vcl.
  [elro]


1.2 (2011-01-11)
----------------

- Added new options ``cache-type``, ``cache-location`` for specifying type of
  Varnish storage (such as using malloc for alternative storage) and setting a
  custom location for said storage
  [davidjb]

- Added additional unit tests to check Varnish initialisation script
  [davidjb]

- Added new option 'purge-hosts'. Enables additional addresses allowed to purge.
  [jensens]

- Added the `name` option to be able to define the directory varnishd
  puts temporary files to and identify the instance when using varnishlog
  or varnishstat.
  [fRiSi]

- fixed configuration for verbose-headers=on (context in vlc_fetch is
  bresp instead of obj in newer varnish versions)
  [fRiSi]

1.1 (2010-08-05)
----------------

- Changed the default cache size to 256M from 1G.
  [hannosch]

- Updated Varnish to 2.1.3.
  [hannosch]

1.1b1 (2010-04-25)
------------------

- Updated advertised Varnish version to 2.1 and adjusted config.
  [hannosch]

- Correct documentation for the ``daemon`` setting and remove the default.
  [hannosch]

- Removed the deprecated build recipe.
  [hannosch]

- Added basic test infrastructure and a test for the simple buildout.
  [hannosch]

- Use the built-in set type instead of the deprecated sets module. This recipe
  now requires at least Python 2.4.
  [hannosch]

- Added the ability to configure runtime parameters in the varnish runner
  configuration and added information to the documentation for it.
  [benliles]

- Improve readability of the generated config.
  [ldr]

1.0.2 (2010-01-18)
------------------

- Update proposed Varnish to 2.0.6.
  [hannosch]

- Further documentation cleanup.
  [hannosch, vincentfretin]

1.0.1 (2009-11-27)
------------------

- Expose the ``download-url`` of a known-good Varnish release that works with
  the configuration produced by the instance recipe.
  [hannosch]

- Consistently use tabs in the generated vcl file.
  [hannosch]

- Whitespace and documentation cleanup.
  [hannosch]


1.0 (2009-08-27)
----------------

* Made the vcl template build its acl purge section. At present, the vcl will
  only allow purges coming from the local host. If we have multiple hosts that
  are separate from localhost, any PURGE requests will be denied without this.
  See http://varnish.projects.linpro.no/wiki/VCLExamplePurging
  [rockdj]

* Added ability to set various Varnish timeouts (connect_timeout,
  first_byte_timeout, and between_bytes_timeout) from each option in the
  buildout. Default values are set at Varnish defaults of 0.4s for
  connect_timeout, and 60s for between_bytes_timeout. Time for
  first_byte_timeout is set at 300s as per plone.recipe.varnish 1.0rc9.
  [rockdj]

* Set `req.http.host` for incoming virtual hosted URLs. Without setting this,
  purge requests sent from hosts other than localhost (the only host in the acl
  purge list) will result in a 404 message. See
  http://davidjb.com/blog/2009/01/plone-varnish-configuration-cache-hits-purge-fails
  [rockdj]


1.0rc11 (2009-06-27)
--------------------

* Reintroduced grace options. What the varnish documentation say about grace:

plone/recipe/varnish/recipe.py
# -*- coding: utf-8 -*-
from plone.recipe.varnish.vclgen import VclGenerator
from zc.recipe.cmmi import Recipe as CMMIRecipe
from zc.recipe.cmmi import system
import logging
import os
import re
import zc.buildout

DEFAULT_DOWNLOAD_URLS = {
    '4.0': 'https://repo.varnish-cache.org/source/varnish-4.0.3.tar.gz',
    '4.1': 'https://repo.varnish-cache.org/source/varnish-4.1.2.tar.gz',
    '4':   'https://repo.varnish-cache.org/source/varnish-4.1.2.tar.gz',
}
DEFAULT_VERSION = '4'

COOKIE_WHITELIST_DEFAULT = """\
statusmessages
__ac
_ZopeId
__cp
"""

COOKIE_PASS_DEFAULT = """\
"__ac(|_(name|password|persistent))=":"\.(js|css|kss)"
"""
COOKIE_PASS_RE = re.compile('"(.*)":"(.*)"')


class BaseRecipe(object):

    def __init__(self, buildout, name, options):
        self.name = name
        self.options = options
        self.buildout = buildout
        self.logger = logging.getLogger(self.name)

    def _version_check(self):
        """check if version is supported
        """
        if self.options['varnish_version'] not in DEFAULT_DOWNLOAD_URLS:
            self._log_and_raise(
                'Varnish {0} is not supported.'.format(
                    self.options['varnish_version']
                )
            )

    def _log_and_raise(self, message):
        """log error first and then raise buildout Exception
        """
        self.logger.error(message)
        raise zc.buildout.UserError(message)

    def _process_bind(self):
        self.options['bind'] = self.options.get('bind').lstrip(':')
        bind = self.options['bind'].split(':')
        if len(bind) == 1 and bind[0].isdigit():
            self.options['bind-host'] = ''
            self.options['bind-port'] = bind[0]
            self.options['bind'] = ':' + bind[0]
        elif len(bind) == 2 and bind[1].isdigit():
            self.options['bind-host'] = bind[0]
            self.options['bind-port'] = bind[1]
        else:
            self._log_and_raise('Invalid syntax for bind')

    def get_from_section(self, part, key, default):
        if part not in self.buildout:
            return default
        if key not in self.buildout[part]:
            return default
        return self.buildout[part][key]

    def install(self):
        pass

    def update(self):
        pass


class BuildRecipe(CMMIRecipe, BaseRecipe):

    def __init__(self, buildout, name, options):
        BaseRecipe.__init__(self, buildout, name, options)
        self.options.setdefault('varnish_version', DEFAULT_VERSION)
        self._version_check()
        self.options.setdefault(
            'url',
            DEFAULT_DOWNLOAD_URLS[options.get('varnish_version')]
        )
        self.options.setdefault('jobs', '4')
        CMMIRecipe.__init__(self, buildout, name, self.options)

    def cmmi(self, dest):
        """Do the 'configure; make; make install' command sequence.

        When this is called, the current working directory is the
        source directory.  The 'dest' parameter specifies the
        installation prefix.

        This is overidden in order to enable parallel jobs in make.
        """
        options = self.configure_options
        if options is None:
            options = '--prefix="%s"' % dest
        if self.extra_options:
            options += ' %s' % self.extra_options

        # C
        system("%s %s" % (self.configure_cmd, options))

        # M
        base_make = 'make'
        if int(self.options.get('jobs')) > 1:
            base_make += ' -j {0}'.format(self.options.get('jobs'))
        system(base_make)

        # MI
        system("make install")

        # TODO: future task: integrate vmods

        # set daemon location
        self.options['daemon'] = os.path.join(
            self.options['location'],
            'sbin',
            'varnishd'
        )


class ConfigureRecipe(BaseRecipe):

    def __init__(self, buildout, name, options):
        super(ConfigureRecipe, self).__init__(buildout, name, options)

        self.options.setdefault('build-part', 'varnish-build')
        self.options.setdefault(
            'varnish_version',
            self.get_from_section(
                self.options['build-part'],
                'varnish_version',
                DEFAULT_VERSION
            )
        )
        self._version_check()

        self.options.setdefault(
            'location',
            os.path.join(buildout['buildout']['parts-directory'], self.name)
        )
        self.options.setdefault('verbose-headers', 'off')
        self.options.setdefault('saint-mode', 'off')
        self.options.setdefault('balancer', 'none')
        self.options.setdefault('backends', '127.0.0.1:8080')
        self.options.setdefault(
            'config-file',
            os.path.join(self.options['location'], 'varnish.vcl')
        )
        self.options.setdefault('connect-timeout', '0.4s')
        self.options.setdefault('first-byte-timeout', '300s')
        self.options.setdefault('between-bytes-timeout', '60s')
        self.options.setdefault('purge-hosts', '')
        self.options.setdefault('cookie-pass', COOKIE_PASS_DEFAULT)
        self.options.setdefault('cookie-whitelist', COOKIE_WHITELIST_DEFAULT)
        # set and test for valid bind value
        self.options.setdefault('bind', '127.0.0.1:8000')
        self._process_bind()

    def _process_backends(self):
        result = []
        raw_backends = [
            _.rsplit(':', 2)
            for _ in self.options['backends'].strip().split()
        ]
        # consistency checks
        if len(raw_backends) > 1:
            lengths = set([len(x) for x in raw_backends])
            if lengths != set([3]):
                self._log_and_raise(
                    'When using multiple backends a hostname '
                    'must be given for each client'
                )
        for idx, raw_backend in enumerate(raw_backends):
            backend = {
                'name': 'backend_{0:03d}'.format(idx)

            }
            try:
                if len(raw_backend) == 3:
                    url, host, port = raw_backend
                else:
                    host, port = raw_backend
                    url = None
            except ValueError:
                self._log_and_raise(
                    'Invalid syntax for backend: {0}'.format(
                        ':'.join(raw_backend)
                    )
                )
                raise zc.buildout.UserError('Invalid syntax for backends')
            backend['url'] = url
            backend['host'] = host
            backend['port'] = port
            backend['connect_timeout'] = self.options['connect-timeout']
            backend['first_byte_timeout'] = self.options['first-byte-timeout']
            backend['between_bytes_timeout'] = \
                self.options['between-bytes-timeout']

            result.append(backend)

        return result

    def _process_zope_vhm_map(self, backends):
        result = {}

        vhm_external_port = self.options.get(
            'zope2_vhm_port',
            self.options['bind-port']
        )
        vhm_proto = 'http'
        if self.options.get('zope2_vhm_ssl', False) == 'on':
            vhm_external_port = self.options.get('zope2_vhm_ssl_port', '443')
            vhm_proto = 'https'

        for line in self.options.get('zope2_vhm_map', '').split():
            domain, location = line.split(':')
            result[domain.strip()] = {
                'location': location.strip(),
                'proto': vhm_proto,
                'external_port': vhm_external_port,
            }

        # consistency checks
        if result:
            lengths = set([len(x) for x in backends])
            if lengths != set([3]):
                self._log_and_raise(
                    'When using VHM a hostname must be given for each backend'
                )
        return result

    def _process_balancers(self, balancer, backends):
        """if theres is a balancer configured, all backends are assigned.

        this could be refactored in future to support multiple balancers.
        """
        result = []
        if balancer != 'none':
            record = {
                'type': balancer,
                'name': 'balancer_0',
                'backends': [_['name'] for _ in backends],
            }
            result.append(record)
        return result

    def install(self):
        if 'configuration-file' not in self.options:
            if not os.path.exists(self.options['location']):
                os.mkdir(self.options['location'])
                self.options.created(self.options['location'])
            self.options['configuration-file'] = os.path.join(
                self.options['location'],
                'varnish.vcl'
            )
        self.create_varnish_configuration()
        return self.options.created()

    def update(self):
        self.install()

    def create_varnish_configuration(self):
        major_version = self.options['varnish_version']
        config = {}
        config['version'] = major_version

        # enable verbose varnish headers
        config['verbose'] = self.options['verbose-headers'] == 'on'

        # Deprecated; see https://www.varnish-cache.org/forum/topic/2777
        # config['saint'] = self.options['saint-mode'] == 'on'
        # if config['saint'] and config['verbose']:
        #     self._log_and_raise(
        #         'When using saint-mode verbose headers must be off'
        #     )

        config['gracehealthy'] = self.options.get('grace-healthy', None)
        config['gracesick'] = self.options.get('grace-sick', 600)

        # fixup cookies for better plone caching
        config['cookiewhitelist'] = [
            _.strip() for _ in self.options['cookie-whitelist'].split()
        ]
        config['cookiepass'] = []
        for line in self.options['cookie-pass'].split():
            line = line.strip()
            if not line:
                continue
            match = COOKIE_PASS_RE.match(line)
            mg = match.groups()
            if not mg and len(mg) != 2:
                continue
            config['cookiepass'].append(
                dict(zip(('match', 'exclude'), mg))
            )
        # inject custom vcl
        config['custom'] = {}
        for name in ('vcl_recv', 'vcl_hit', 'vcl_miss', 'vcl_fetch',
                     'vcl_deliver', 'vcl_pipe'):
            config['custom'][name] = self.options.get(name, '')

        config['backends'] = self._process_backends()
        config['directors'] = self._process_balancers(
            self.options['balancer'].strip(),
            config['backends']
        )
        config['zope2_vhm_map'] = self._process_zope_vhm_map(
            config['backends']
        )
        config['code404page'] = True

        # build the purge host string
        config['purgehosts'] = set([])
        for segment in self.options['purge-hosts'].split():
            segment = segment.strip()
            if segment:
                config['purgehosts'].add(segment)

        vclgenerator = VclGenerator(config)
        filedata = vclgenerator()
        with open(self.options['configuration-file'], 'wt') as fio:
            fio.write(filedata)
        self.options.created(self.options['configuration-file'])


class ScriptRecipe(BaseRecipe):

    def __init__(self, buildout, name, options):
        super(ScriptRecipe, self).__init__(buildout, name, options)

        self.options.setdefault('build-part', 'varnish-build')
        self.options.setdefault('configuration-part', 'varnish-configuration')

        self.options.setdefault(
            'daemon',
            self.get_from_section(
                self.options['build-part'],
                'location',
                '/usr'
            ) + '/sbin/varnishd'
        )

        self.options.setdefault(
            'bind',
            self.get_from_section(
                self.options['configuration-part'],
                'bind',
                '127.0.0.1:8000'
            )
        )
        self.options.setdefault(
            'configuration-file',
            self.get_from_section(
                self.options['configuration-part'],
                'config-file',
                ''
            )
        )
        if not self.options['configuration-file']:
            self._log_and_raise('No configuration file found')
        self._process_bind()

        self.options.setdefault(
            'location',
            os.path.join(buildout['buildout']['parts-directory'], self.name)
        )
        self.options.setdefault('cache-type', 'file')
        self.options.setdefault('cache-size', '256M')
        self.options.setdefault('runtime-parameters', '')
        self.options.setdefault('secret-file', 'nosecret')

    def install(self):
        if 'cache-location' not in self.options:
            if not os.path.exists(self.options['location']):
                os.mkdir(self.options['location'])
                self.options.created(self.options['location'])
            self.options['cache-location'] = os.path.join(
                self.options['location'],
                'storage'
            )
            if not os.path.exists(self.options['cache-location']):
                os.mkdir(self.options['cache-location'])
                self.options.created(self.options['cache-location'])

        self.create_varnish_script()
        return self.options.created()

    def create_varnish_script(self):
        target = os.path.join(
            self.buildout['buildout']['bin-directory'],
            self.name
        )
        parameters = self.options['runtime-parameters'].strip().split()

        with open(target, 'wt') as tf:
            # XXX TODO: refactor me! make this a template.
            print >>tf, '#!/bin/sh'
            print >>tf, 'exec %s \\' % self.options['daemon']
            if 'user' in self.options:
                print >>tf, '    -p user=%s \\' % self.options['user']
            if 'group' in self.options:
                print >>tf, '    -p group=%s \\' % self.options['group']
            print >>tf, '    -f "%s" \\' % self.options['configuration-file']
            print >>tf, '    -P "%s" \\' % \
                os.path.join(self.options['location'], 'varnish.pid')
            print >>tf, '    -a %s \\' % self.options['bind']
            if self.options.get('telnet', None):
                print >>tf, '    -T %s \\' % self.options['telnet']
            if self.options['cache-type'] == 'malloc':
                print >>tf, '    -s %s,%s \\' % (
                    self.options['cache-type'],
                    self.options['cache-size']
                )
            else:
                print >>tf, '    -s %s,"%s",%s \\' % (
                    self.options['cache-type'],
                    self.options['cache-location'],
                    self.options['cache-size']
                )
            if self.options.get('mode', 'daemon') == 'foreground':
                print >>tf, '    -F \\'
            if self.options.get('name', None):
                print >>tf, '    -n %s \\' % self.options['name']
            if not self.options.get('secret-file', 'nosecret') == 'nosecret':
                if self.options['secret-file'].lower() == "disabled":
                    # disable authentication on admin interface, dangerous
                    print >>tf, '    -S \\'
                else:
                    # use shared secret file for admin auth
                    print >>tf, '    -S %s \\' % self.options['secret-file']
            for parameter in parameters:
                print >>tf, '    -p %s \\' % (parameter)
            print >>tf, '    "$@"'
        os.chmod(target, 0755)
        self.options.created(target)
